javalib/src/main/scala/java/util/regex/Matcher.scala

```diff
@@ -275,5 +283,358 @@ object Matcher {
+    val allMatchingRegex = new js.RegExp(allMatchingRegex_s, flags)
+    allMatchingRegex.lastIndex = start()
+    val allMatchResult = allMatchingRegex.exec(string)
+    if (allMatchResult == null) {
+      throw new Exception(s"[Internal error] Executed '$allMatchingRegex' on " +
+           s"'$string' at position $start, got an error.\n" +
+           s"Original pattern '${regex}' did match however.")
+    }
+    node.setMatch((x: Int) => allMatchResult(x).getOrElse(null))
+    node.setStartReturnEnd(start())
+    mapping.mapValues(_.start)
+  }
+
+  // parentheses aspect of the regex
+  private[regex] sealed trait OriginalRegex {
+    @inline def fold[A](numberF: Int => A, leftRightF: (String, String) => A): A = this match {
+      case OriginallyGroupped(number) => numberF(number)
+      case OriginallyWrapped(left, right) => leftRightF(left, right)
+    }
+  }
+  // If the regex had originally a group number, that is, with simple parentheses
+  // non-grouping parentheses (?: ...) are always added
+  private[regex] case class OriginallyGroupped(number: Int) extends OriginalRegex
+  // If the regex was not originally groupped with parentheses,
+  // then right holds the left and right modifiers (e.g. *, +, *?, +?, ?)
+  private[regex] case class OriginallyWrapped(left: String, right: String) extends OriginalRegex
+
+  private[regex] object OriginallyWrapped {
+    @inline def isRepeatModifier(modifier: String): Boolean =
+      modifier == "?" || modifier == "??" || modifier == "*" || modifier == "+" ||
+      modifier == "*?" || modifier == "+?" || modifier.startsWith("{")
+
+    object Repeater {
+      def unapply(e: OriginallyWrapped): Option[String] =
+        if (e.left == "" && isRepeatModifier(e.right)) Some(e.right) else None
+    }
+    object Absolute {
+      def unapply(e: OriginallyWrapped): Boolean =
+        (e.left == "(?!" || e.left == "(?=") && e.right == ")"
+    }
+  }
+
+  // Type of the node, a regex or a sequence of nodes.
+  private[regex] sealed trait NodeType {
+    @inline def fold[A](regexF: String => A, childrenF: Seq[Node] => A): A = this match {
+      case RegexLeaf(regex) => regexF(regex)
+      case ParentNode(children) => childrenF(children)
+    }
+  }
+  // A pure regex, without any subgroups
+  private[regex] case class RegexLeaf(regex: String) extends NodeType
+  // A parent groupping, with children nodes.
+  // Modifiers to this group, such as star or plus, can be added using
+  // the field of Node originalGroup.
+  private[regex] case class ParentNode(children: Seq[Node]) extends NodeType
+
+  /** @param originalGroup Either the 1-based index of the original group that
+             this node encloses, or the modifiers which are present before and
+             after (ex: ?, *?, (?=...), (?!..))
+    * @param nodeType The type of the node (Regexleaf, or ParentNode)
+    */
+  private[regex] case class Node(originalGroup: OriginalRegex, var nodeType: NodeType) {
```

Comment:
Added a comment. "Is a var because we need to shift the backreference groups".
Node is a case class for the sake of easy construction and debugging.

Reply:
If it's only for easy construction and debugging, it is preferable to define a `def apply` in the companion object, and override `toString()`.

Reply:
Did so.

Reply:
Actually, I also need to deconstruct the Node for pattern matching. Is it still worth having a object with `apply/unapply` and the `toString` override? I thought that case classes were doing exactly this.

Reply:
Hum, I see. Well, it's up to you. Whichever you prefer. Probably at this point the case class was a better idea, although I'd call it a code smell if we unapply mutable data structures. But since this is confined in a private area, we need not be that strict.